allennlp/data/token_indexers/wordpiece_indexer.py

```diff
@@ -137,39 +175,53 @@ def tokens_to_indices(self,
+            next_offset = 1 if self.use_starting_offsets else 0
+            if self._truncate_long_sequences and offset >= window_length + next_offset:
                 break
 
-        # By construction, we still have enough room to add the end_token ids.
-        wordpiece_ids.extend(self._end_piece_ids)
+            # For initial offsets, the current value of ``offset`` is the start of
+            # the current wordpiece, so add it to ``offsets`` and then increment it.
+            if self.use_starting_offsets:
+                offsets.append(offset)
+                offset += len(token)
+            # For final offsets, the current value of ``offset`` is the end of
+            # the previous wordpiece, so increment it and then add it to ``offsets``.
+            else:
+                offset += len(token)
+                offsets.append(offset)
+
+        if len(flat_wordpiece_ids) <= window_length:
+            # If all the wordpieces fit, then we don't need to do anything special
+            wordpiece_windows = [self._start_piece_ids + flat_wordpiece_ids + self._end_piece_ids]
+        elif self._truncate_long_sequences:
+            logger.warning("Too many wordpieces, truncating sequence. If you would like a sliding window, set"
+                           "`truncate_long_sequences` to False %s", str([token.text for token in tokens]))
+            wordpiece_windows = [self._start_piece_ids + flat_wordpiece_ids[:window_length] + self._end_piece_ids]
+        else:
+            # Create a sliding window of wordpieces of length `max_pieces` that advances by `stride` steps and
+            # add start/end wordpieces to each window
+            # TODO: this currently does not respect word boundaries, so words may be cut in half between windows
+            # However, this would increase complexity, as sequences would need to be padded/unpadded in the middle
+            wordpiece_windows = [self._start_piece_ids +
+                                 flat_wordpiece_ids[i:i + window_length] +
+                                 self._end_piece_ids
+                                 for i in range(0, len(flat_wordpiece_ids), stride)]
+
+            # Check for overlap in the last window. Throw it away if it is redundant.
+            last_window = wordpiece_windows[-1][1:]
+            penultimate_window = wordpiece_windows[-2]
+            if last_window == penultimate_window[-len(last_window):]:
+                wordpiece_windows = wordpiece_windows[:-1]
+
+        # Flatten the wordpiece windows
+        wordpiece_ids = [wordpiece for sequence in wordpiece_windows for wordpiece in sequence]
+
         # Constructing `token_type_ids` by `self._separator`
-        token_type_ids = _get_token_type_ids(wordpiece_ids,
-                                             self._separator_ids)
+        token_type_ids = _get_token_type_ids(wordpiece_ids, self._separator_ids)
```

Comment:
I feel like this is wrong? Imagine I have a "[SEP]" token between my sentences, it could end up in multiple windows? Then the flattened wordpiece_ids will have multiple "[SEP]" tokens, which means this will be doing the wrong thing. Am I missing something?

Reply:
I just pushed up a test that fails because of this, do you want me to take a stab at fixing it?

Reply:
ok, I pushed up what should be a fix